controllers/admin/AdminProductsController.php

```diff
@@ -3214,4 +3214,157 @@ public function processImageLegends()
             }
         }
     }
+
+    /**
+     * Returns in an homemade JSON with the content of a products pack.
+     */
+    public function displayAjaxProductPackItems()
+    {
+        $jsonArray = array();
+        $products = Db::getInstance()->executeS('
+            SELECT p.`id_product`, pl.`name`
+            FROM `' . _DB_PREFIX_ . 'product` p
+            NATURAL LEFT JOIN `' . _DB_PREFIX_ . 'product_lang` pl
+            WHERE pl.`id_lang` = ' . (int) (Tools::getValue('id_lang')) . '
+            ' . Shop::addSqlRestrictionOnLang('pl') . '
+            AND NOT EXISTS (SELECT 1 FROM `' . _DB_PREFIX_ . 'pack` WHERE `id_product_pack` = p.`id_product`)
+            AND p.`id_product` != ' . (int) (Tools::getValue('id_product')));
+
+        foreach ($products as $packItem) {
+            $jsonArray[] = '{"value": "' . (int) ($packItem['id_product']) . '-' . addslashes($packItem['name']) . '", "text":"' . (int) ($packItem['id_product']) . ' - ' . addslashes($packItem['name']) . '"}';
```

Comment:
This looks like a good target for a refacto into a private function

Reply:
What would you see in this private method?